library/core/src/num/mod.rs

```diff
@@ -846,37 +866,40 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize {
+        // SAFETY: We can take this fast path when `radix.pow(digits.len()) - 1 <= T::MAX`
+        // but the condition above is a faster (conservative) approximation of this.
+        //
+        // Consider the highest radix of 16:
+        // `u8::MAX` is `ff` (2 characters), `u16::MAX` is `ffff` (4 characters)
+        // We can be sure that any src len of 2 would fit in a u8 so we don't need
+        // to check for overflow.
+        unsafe {
+            let unchecked_additive_op =
+                if is_positive { T::unchecked_add } else { T::unchecked_sub };
+
+            for &c in digits {
+                result = result.unchecked_mul(radix);
+                let x = (c as char).to_digit(radix).ok_or(PIE { kind: InvalidDigit })?;
+                result = unchecked_additive_op(&result, x);
+            }
         }
     } else {
-        // The number is negative
+        let additive_op = if is_positive { T::checked_add } else { T::checked_sub };
+        let overflow_err = || PIE { kind: if is_positive { PosOverflow } else { NegOverflow } };
+
         for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: NegOverflow }),
-            };
-            result = match result.checked_sub(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: NegOverflow }),
-            };
+            // When `radix` is passed in as a literal, rather than doing a slow `imul`
+            // the compiler can use shifts if `radix` can be expressed as a
+            // sum of powers of 2 (x*10 can be written as x*8 + x*2).
+            // When the compiler can't use these optimisations,
+            // the latency of the multiplication can be hidden by issuing it
+            // before the result is needed.
```

Comment:
```suggestion
            // When the compiler can't use these optimisations,
            // the latency of the multiplication can be hidden by issuing it
            // before the result is needed to improve performance on
            // modern out-of-order CPU as multiplication here is slower
            // than the other instructions, we can get the end result faster
            // doing multiplication first and let the CPU spends other cycles
            // doing other computation and get multiplication result later.
```

I really didn't know about the internals of out-of-order CPU, maybe we can explain a bit?

Reply:
I am all for this. The std lib unlike the compiler source code is read by a lot of people and the more it can be used to provide insights of what's going on behind the scenes the better. This is why I was keen on being explicit about what the compiler is doing with *8 and *10 because someone that didn't know all the crazy things a compiler can do could look at that code and try to optimise it and waste effort.

Reply:
(I found out yesterday from Jane that if you did not care about looking at the type of the parse error and mapped it away: `parse(...).err_map(|_| ())` then you get better machine code because it only has to encode that there was an error, not what the error was. Was totally astounded that it could be that clever.)